fish_prompt.fish

```diff
@@ -4,11 +4,170 @@
 # MIT License
 
 function fish_prompt
-  set -l prompt ""
-
   # Save previous exit code
   set -l exit_code $status
 
+  # Template
+  set -l user[3] ""
+  set -l at[3] ""
+  set -l host[3] ""
+  set -l git_branch[3] ""
+  set -l git_dirty[3] ""
+  set -l git_arrows[3] ""
+  set -l command_duration[3] ""
+  set -l current_folder[3] ""
+
+  set -l prefix ""
+  # Do not add a line break to a brand new session
+  if test $__pure_fresh_session -eq 0
+    set prefix "\n"
+  end
+
+  # Clear the line
+  set prefix $prefix "\r\033[K"
+
+  # Check if user is in an SSH session
+  if [ "$SSH_CONNECTION" != "" ]
```

Comment:
Is there any reason to use `source` instead of placing functions inside `functions` folder? We do have `__parse_cutrent_forlder` and other functions there.

I’m also wondering if we should prefix them with `pure` or something similar, since those are available everywhere in fish. How we can sure it won’t collide with other user packages? Is it job for the package manager or there is a naming convention we should adopt?

> I will advise not to use the array approach
The problem is we need to calculate string length later, to do that reliably we need to separate text from styling commands, since they are count as symbols when measuring string. If you know how to measure actual number of symbols displayed on the terminal instead of string length, that would solve the problem.

Reply:
### `functions/`
Using `functions/` is fine, I forgot about [autoload mechanism](https://fishshell.com/docs/current/tutorial.html#tut_autoload) and I didn't think where to put those new files :+1: 

### Prefixing function name

Indeed, I think we should prefix our function's names to prevent conflict, maybe  @jorgebucaran can give us some insight on this. 
I created issue #90 for that, we should do this in another PR as it will impact functions unrelated to this issue and tests.

### calculate string length

Could you elaborate on number of symbols vs. string length and provide some example of what we have and what should be the result?

Reply:
The internal `string length` seem to compute corectly:

```fish
❯ set -l down_arrow "⇣"
❯ string length $down_arrow
1

❯ set -l down_arrow_red (set_color red)"⇣"(set_color normal)
❯ string length $down_arrow_red
15
```

Reply:
> Could you elaborate on number of symbols vs. string length and provide some example

You've provided perfect example yourself: `$down_arrow_red` is 15 long string, but it prints only one symbol in terminal.

We need to count the number of _actual_ symbols that will be printed in terminal and compare it with number of columns in terminal window, so our prompt won't exceed it.

Reply:
I just ask the question on stackoverflow: [How to count symbol instead of string length with Fish?](https://stackoverflow.com/q/53666154/802365).

While writing it I thought about using `wc`, tell me if that will work:

```fish
❯ set -l down_arrow_red (set_color red)"⇣"(set_color normal)
❯ echo $down_arrow_red | wc -w
1
```

Reply:
Well, it seems, filter out the color sequences with regular expressions using `string replace` like proposed [here](https://github.com/fish-shell/fish-shell/issues/5390#issuecomment-445286812) is a way to go, at least until we don't have [`string width`](https://github.com/fish-shell/fish-shell/issues/4012) implemented in fish.

We don't need to use array approach at all then.